pkgs/test_api/test/backend/invoker_test.dart

```diff
@@ -449,49 +449,67 @@ void main() {
     });
   });
 
-  group('waitForOutstandingCallbacks:', () {
-    test('waits for the wrapped function to complete', () async {
-      var functionCompleted = false;
-      await Invoker.current!.waitForOutstandingCallbacks(() async {
-        await pumpEventQueue();
-        functionCompleted = true;
-      });
-
-      expect(functionCompleted, isTrue);
+  group('runTearDowns', () {
+    test('runs multiple tear downs', () async {
+      var firstTearDownStarted = false;
+      var secondTearDownStarted = false;
+      await Invoker.current!.runTearDowns([
+        () {
+          firstTearDownStarted = true;
+        },
+        () {
+          secondTearDownStarted = true;
+        }
+      ]);
+      expect(secondTearDownStarted, isTrue);
+      expect(firstTearDownStarted, isTrue);
     });
 
-    test('waits for registered callbacks in the wrapped function to run',
+    test('waits for the future returned tear downs to complete', () async {
+      var firstTearDownWork = Completer<void>();
+      var secondTearDownStarted = false;
+      var result = Invoker.current!.runTearDowns([
+        () {
+          secondTearDownStarted = true;
+        },
         () async {
-      var callbackRun = false;
-      await Invoker.current!.waitForOutstandingCallbacks(() {
-        pumpEventQueue().then(expectAsync1((_) {
-          callbackRun = true;
-        }));
-      });
-
-      expect(callbackRun, isTrue);
+          await firstTearDownWork.future;
+        },
+      ]);
+      await pumpEventQueue();
+      expect(secondTearDownStarted, isFalse);
+      firstTearDownWork.complete();
+      await result;
+      expect(secondTearDownStarted, isTrue);
     });
 
-    test("doesn't automatically block the enclosing context", () async {
-      var innerFunctionCompleted = false;
-      await Invoker.current!.waitForOutstandingCallbacks(() {
-        Invoker.current!.waitForOutstandingCallbacks(() async {
-          await pumpEventQueue();
-          innerFunctionCompleted = true;
-        });
-      });
-
-      expect(innerFunctionCompleted, isFalse);
+    test('allows next tear down to run while there are still prior callbacks',
```

Comment:
I don't follow how this is possible. It looks like we don't complete until the teardown future completes in `invoker.dart` line `236`. Care to explain?

Reply:
This test corresponds to this bullet point from the commit message:

> * Tear downs do not need to wait for all outstanding callbacks from
>   previous tear downs before they can start. They do wait for the
>   `Future` returned from prior tear downs. This is the reason that the
>   loop waits for the `Completer` which is completed following the tear
>   down callback, and the remainder of the `waitForOutstandingCallbacks`
>   work is `unawaited`.

To elaborate - the `unawaited` is ignoring the work done by `_waitForOutstandingCallbacks`, whereas the `Completer` is capturing the work encompassed by the future that is returned by the callback. So if you `await` stuff in your `tearDown` callback and you are returning a future, that does cause the other teardowns to wait. However if you `expect(someOtherFuture, completes)` then that `someOtherFuture` is an outstanding callback so the overall test still wants it to complete before it can be considered done, but other teardowns can get started.